tests/garage/tf/core/test_cnn.py

```diff
@@ -0,0 +1,226 @@
+        self.batch_size = 5
+        self.input_width = 10
+        self.input_height = 10
+        self.obs_input = np.ones((self.batch_size, self.input_width,
+                                  self.input_height, 3))
+
+        input_shape = self.obs_input.shape[1:]  # height, width, channel
+        self._input_ph = tf.placeholder(
+            tf.float32, shape=(None, ) + input_shape, name="input")
+
+        self._output_shape = 2
+        self.filter_sizes = (3, )
+        self.in_channels = (3, )
+        self.out_channels = (32, )
+
+        self.hidden_nonlinearity = tf.nn.relu
+
+    def test_output_shape(self):
+
+        with tf.variable_scope("CNN"):
+            self.cnn = cnn(
+                input_var=self._input_ph,
+                output_dim=self._output_shape,
+                filter_dims=self.filter_sizes,
+                num_filters=self.out_channels,
+                stride=1,
+                name="cnn1",
+                padding="VALID",
+                hidden_w_init=tf.constant_initializer(1),
+                hidden_nonlinearity=self.hidden_nonlinearity)
+
+        self.sess.run(tf.global_variables_initializer())
+
+        result = self.sess.run(
+            self.cnn, feed_dict={self._input_ph: self.obs_input})
+        assert result.shape[1] == self._output_shape
+
+    def test_output_with_identity_filter(self):
+
+        with tf.variable_scope("CNN"):
+            self.cnn = cnn(
+                input_var=self._input_ph,
+                output_dim=self._output_shape,
+                filter_dims=self.filter_sizes,
+                num_filters=self.out_channels,
+                stride=1,
+                name="cnn1",
+                padding="VALID",
+                hidden_w_init=tf.constant_initializer(1),
+                hidden_nonlinearity=self.hidden_nonlinearity)
+
+        self.sess.run(tf.global_variables_initializer())
+
+        result = self.sess.run(
+            self.cnn, feed_dict={self._input_ph: self.obs_input})
+
+        # get weight values
+        with tf.variable_scope("CNN", reuse=True):
+            out_w = tf.get_variable("cnn1/output/kernel")
+            out_b = tf.get_variable("cnn1/output/bias")
+
+        filter_sum = 1
+        # filter value after 3 layers of conv
+        for filter_size, in_channel in zip(self.filter_sizes,
+                                           self.in_channels):
+            filter_sum *= filter_size * filter_size * in_channel
+
+        # input shape 10 * 10 * 3
+        # after conv layer, we have filter with shape 8 * 8 * 32 (last channel)
+        # flatten
+        h_out = np.full(
+            (self.batch_size, 8 * 8 * self.out_channels[-1]),
+            filter_sum,
+            dtype=np.float32)
+        # pass to a dense layer
+        dense_in = tf.matmul(h_out, out_w) + out_b
+        np.testing.assert_array_equal(dense_in.eval(), result)
+
+    def convolve(self, input, filter_weights, filter_biass, stride):
```

Comment:
Please use a different name for `input`.

Reply:
Any specific reason? In the generic case `input` is a bad name for a function parameter, but in this case `convolve` is a mathematical function of `input`, and it's helpful to know which parameter is the "input" to the function being represented.

Reply:
`input()` is a [Python builtin](https://docs.python.org/3/library/functions.html), so using the variable name `input` masks the symbol in this context.

PEP8 [gently recommends](https://www.python.org/dev/peps/pep-0008/#descriptive-naming-styles) using a trailing underscore to avoid overloading builtins. I've also used `an_input` in the past.